lib/pbench/server/api/resources/query_apis/metadata_index/__init__.py

```diff
@@ -0,0 +1,103 @@
+from http import HTTPStatus
+from logging import Logger
+from typing import AnyStr
+
+from flask_restful import abort
+
+from pbench.server import PbenchServerConfig
+from pbench.server.api.resources import JSON, Schema, SchemaError
+from pbench.server.api.resources.query_apis import CONTEXT, ElasticBase
+from pbench.server.database.models.datasets import Dataset, Metadata, MetadataError
+from pbench.server.database.models.users import User
+
+
+class MissingRunIdSchemaParameter(SchemaError):
+    """
+    The subclass schema is missing the required "run_id" parameter required
+    to locate a Dataset.
+    """
+
+    def __init__(self, subclass_name: str):
+        super().__init__()
+        self.subclass_name = subclass_name
+
+    def __str__(self) -> str:
+        return f"API {self.subclass_name} is missing schema parameter run_id"
+
+
+class RunIdBase(ElasticBase):
+    """
+    A base class for query apis that depends on Metadata for getting the
+    indices.
+
+    This class extends the ElasticBase class and implements a common
+    `preprocess` method based on client provided run id. The subsequent methods
+    such as 'assemble' and 'postprocess' need to be implemented by a respective
+    subclasses.
+
+    Note that run_id is a required schema parameter for all the
+    classes inheriting this class. Also "preprocess" provides json context
+    containing dataset and a run_id that's passed to the assemble and
+    postprocess methods.
+    """
+
+    def __init__(self, config: PbenchServerConfig, logger: Logger, schema: Schema):
+        if "run_id" not in schema:
+            raise MissingRunIdSchemaParameter(self.__class__.__name__)
+        super().__init__(config, logger, schema)
+
+    def preprocess(self, client_json: JSON) -> CONTEXT:
+        """
+        Query the Dataset associated with this run id, and determine whether the
+        request is authorized for this dataset.
+
+        If the user has authorization to read the dataset, return the Dataset
+        object and run_id as JSON CONTEXT so that the postprocess operations
+        can use it to identify the index to be searched from document index
+        metadata.
+        """
+        run_id = client_json["run_id"]
+
+        # Query the dataset using the given run id
+        dataset = Dataset.query(md5=run_id)
+        if not dataset:
+            self.logger.debug(f"Dataset with Run ID {run_id!r} not found")
```

Comment:
This really should be not using f-strings to avoid always rendering the `repr` of `run_id` even if it is not emitted.

Reply:
This happens in many places, so perhaps a follow-up change to this PR?

Reply:
Sure I will address this in timeseries query api that I am working on